src/PricingModules/UniswapV3/UniswapV3WithFeesPricingModule.sol

```diff
@@ -0,0 +1,442 @@
+/**
+ * Created by Pragma Labs
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+pragma solidity ^0.8.13;
+
+import { PricingModule, IPricingModule } from "../AbstractPricingModule.sol";
+import { IMainRegistry } from "../interfaces/IMainRegistry.sol";
+import { INonfungiblePositionManager } from "./interfaces/INonfungiblePositionManager.sol";
+import { IUniswapV3Pool } from "./interfaces/IUniswapV3Pool.sol";
+import { IERC20 } from "./interfaces/IERC20.sol";
+import { TickMath } from "./libraries/TickMath.sol";
+import { FullMath } from "./libraries/FullMath.sol";
+import { PoolAddress } from "./libraries/PoolAddress.sol";
+import { FixedPoint96 } from "./libraries/FixedPoint96.sol";
+import { FixedPoint128 } from "./libraries/FixedPoint128.sol";
+import { LiquidityAmounts } from "./libraries/LiquidityAmounts.sol";
+import { FixedPointMathLib } from "lib/solmate/src/utils/FixedPointMathLib.sol";
+import { SafeCastLib } from "lib/solmate/src/utils/SafeCastLib.sol";
+
+/**
+ * @title Pricing Module for Uniswap V3 Liquidity Positions.
+ * @author Pragma Labs
+ * @notice The pricing logic and basic information for Uniswap V3 Liquidity Positions.
+ * @dev The UniswapV3PricingModule will not price the LP-tokens via direct price oracles,
+ * it will break down liquidity positions in the underlying tokens (ERC20s).
+ * Only LP tokens for which the underlying tokens are allowed as collateral can be priced.
+ * @dev No end-user should directly interact with the UniswapV3PricingModule, only the Main-registry,
+ * or the contract owner.
+ */
+contract UniswapV3WithFeesPricingModule is PricingModule {
+    using FixedPointMathLib for uint256;
+    using FullMath for uint256;
+
+    /* //////////////////////////////////////////////////////////////
+                                STORAGE
+    ////////////////////////////////////////////////////////////// */
+
+    // Map asset => uniswapV3Factory.
+    mapping(address => address) public assetToV3Factory;
+
+    // The Arcadia Pricing Module for standard ERC20 tokens (the underlying assets).
+    PricingModule immutable erc20PricingModule;
+
+    /* //////////////////////////////////////////////////////////////
+                                CONSTRUCTOR
+    ////////////////////////////////////////////////////////////// */
+
+    /**
+     * @param mainRegistry_ The contract address of the MainRegistry.
+     * @param oracleHub_ The contract address of the OracleHub.
+     * @param riskManager_ The address of the Risk Manager.
```

Comment:
`* @param erc20PricingModule_ The contract address of the Pricing Module for ERC20s`